### DISCUSSION TITLE
Is it possible to create a macOS universal app on M1 chip?

### DISCUSSION THREAD

**Discussion by arossert** [pyinstaller/pyinstaller#6650] [Help]
I'm using PyInstaller to bundle my application on macOS and wanted to know if this is possible to build a universal app on M1 that will also work on Intel chips.

When I build the application on Intel it also works on M1 but if I build on M1 I get an error on Intel:
```
bad CPU type in executable: ./app
```

I'm using the latest PyInstaller4.10 release.

**Comment by bwoodsend** [comment]
Does [what's already written here](https://pyinstaller.readthedocs.io/en/stable/feature-notes.html#macos-multi-arch-support) not do the job?

**Reply by rokm** [reply]
> When I build the application on Intel it also works on M1 but if I build on M1 I get an error on Intel

That's because x86_64 build made on Intel runs on M1 under rosetta2, while arm64 build made on M1 won't run on Intel.

As per docs that Brénainn linked above, PyInstaller by default makes single-arch builds that target the running architecture. You can try making a universal2 build, but it requires universal2 python build as well as *all* python packages that you use to be either pure-python or provide universal2 builds of their extensions (i.e., a single single-arch wheel, such as `numpy`, will make it impossible to create universal2 application).

**Reply by rokm** [reply]
So if you cannot make a universal2 build (due to dependencies), you can still make an x86_64 build on M1 by running a x86_64-only python environment under rosetta2.

**Reply by arossert** [reply]
@bwoodsend I have totally missed that part, thanks.
